core/services/ocr2/plugins/promwrapper/plugin.go

```diff
@@ -80,93 +97,251 @@ var (
+}
+
+func (*defaultPometheusBackend) SetShouldTransmitAcceptedReportDuration(labelValues []string, duration float64) {
+	promShouldTransmitAcceptedReport.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPometheusBackend) SetCloseDuration(labelValues []string, duration float64) {
+	promClose.WithLabelValues(labelValues...).Observe(duration)
+}
+
+func (*defaultPometheusBackend) SetQueryToObservationLatency(labelValues []string, latency float64) {
+	promQueryToObservationLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetObservationToReportLatency(labelValues []string, latency float64) {
+	promObservationToReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetReportToAcceptFinalizedReportLatency(labelValues []string, latency float64) {
+	promReportToAcceptFinalizedReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func (*defaultPometheusBackend) SetAcceptFinalizedReportToTransmitAcceptedReportLatency(labelValues []string, latency float64) {
+	promAcceptFinalizedReportToTransmitAcceptedReportLatency.WithLabelValues(labelValues...).Observe(latency)
+}
+
+func New(
+	plugin types.ReportingPlugin,
+	name string,
+	chainType string,
+	chainID *big.Int,
+	config types.ReportingPluginConfig,
+	backend PrometheusBackend,
+) types.ReportingPlugin {
+	// Apply passed-in Prometheus backend if one is given.
+	var prometheusBackend PrometheusBackend = &defaultPometheusBackend{}
+	if backend != nil {
+		prometheusBackend = backend
+	}
+
 	return &promPlugin{
-		wrapped:      plugin,
-		name:         name,
-		chainType:    chainType,
-		chainID:      chainID,
-		oracleID:     fmt.Sprintf("%d", config.OracleID),
-		configDigest: common.Bytes2Hex(config.ConfigDigest[:]),
+		wrapped:                       plugin,
+		name:                          name,
+		chainType:                     chainType,
+		chainID:                       chainID,
+		oracleID:                      fmt.Sprintf("%d", config.OracleID),
+		configDigest:                  common.Bytes2Hex(config.ConfigDigest[:]),
+		queryEndTimes:                 make(map[types.ReportTimestamp]time.Time),
+		observationEndTimes:           make(map[types.ReportTimestamp]time.Time),
+		reportEndTimes:                make(map[types.ReportTimestamp]time.Time),
+		acceptFinalizedReportEndTimes: make(map[types.ReportTimestamp]time.Time),
+		prometheusBackend:             prometheusBackend,
 	}
 }
 
 func (p *promPlugin) Query(ctx context.Context, timestamp types.ReportTimestamp) (types.Query, error) {
 	start := time.Now().UTC()
 	defer func() {
 		duration := float64(time.Now().UTC().Sub(start))
-		promQuery.WithLabelValues(p.getLabelsValues()...).Observe(duration)
+		p.prometheusBackend.SetQueryDuration(getLabelsValues(p, timestamp), duration)
+		p.queryEndTimes[timestamp] = time.Now() // note time at end of Query()
```

Comment:
I'm indifferent to which timestamp we use but looks like we use `Now().UTC()` above, should we use that here for consistency?

Reply:
Ah yeah thanks. Shouldn't change the current logic but nicer to do that